cmd/snapd-generator/main.c

```diff
@@ -97,3 +91,106 @@ int main(int argc, char **argv)
 	fprintf(f, "WantedBy=local-fs.target\n");
 	return 0;
 }
+
+static bool file_exists(const char *path) {
+	struct stat buf;
+	return stat(path, &buf) == 0;
+}
+
+static bool executable_exists(const char *name) {
+	char *path = getenv("PATH");
+	if (path == NULL) {
+		return false;
+	}
+	char *path_copy SC_CLEANUP(sc_cleanup_string) = NULL;
+	path_copy = sc_strdup(path);
+
+	char *ptr = NULL;
+	char *token = strtok_r(path_copy, ":", &ptr);
+	char fname[PATH_MAX + 1] = { 0 };
+	while (token) {
+		sc_must_snprintf(fname, sizeof fname, "%s/%s", token, name);
+		if (access(fname, X_OK) == 0) {
+			return true;
+		}
+		token = strtok_r(NULL, ":", &ptr);
+	}
+	return false;
+}
+
+int ensure_fusesquashfs_inside_container(const char *normal_dir)
+{
+	// check if we are running inside a container, systemd
+	// provides this file all the way back to trusty if run in a
+	// container
+	if (!file_exists("/run/systemd/container")) {
+		return 0;
+	}
+
+	char *fstype = NULL;
+	if (executable_exists("squashfuse")) {
+		fstype = "fuse.squashfuse";
+	} else if (executable_exists("snapfuse")) {
+		fstype = "fuse.snapfuse";
+	} else {
+		fprintf(stderr,
+			"cannot find squashfuse or snapfuse executable\n");
+		return 2;
+	}
+
+	DIR *units_dir SC_CLEANUP(sc_cleanup_closedir) = NULL;
+	units_dir = opendir("/etc/systemd/system");
+	if (units_dir == NULL) {
+		// nothing to do
+		return 0;
+	}
+
+	char fname[PATH_MAX + 1] = { 0 };
+
+	struct dirent *ent;
+	while (ent = readdir(units_dir)) {
+		if (!(sc_startswith(ent->d_name, "snap-") && sc_endswith(ent->d_name, ".mount"))) {
```

Comment:
You need to handle the `/var/lib/snapd/snap` location as well.

Reply:
Done.